### PR TITLE
Buffered writer for exactly once sink

### PR DESCRIPTION
**Pull request by jayehwhyehentee**
Create writer used by exactly once sink, following the two-phase commit protocol. This writer uses BQ's buffered stream for appending data to the table.

Note to reviewers: this PR is bloated by tests
 
/gcbrun

### CODE REVIEW

**Comment by jayehwhyehentee** [comment]
@clmccart kindly review this PR

**Review by prashastia — APPROVED** [review]

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java [hunk 1]
```diff
@@ -138,24 +150,43 @@ void addToAppendRequest(ByteString protoRow) {
...
     }
 
     /** Creates a StreamWriter for appending to BigQuery table. */
-    StreamWriter createStreamWriter(boolean enableConnectionPool) {
-        logger.debug("Creating BigQuery StreamWriter in subtask {}", subtaskId);
+    void createStreamWriter(boolean enableConnectionPool) {
         try {
-            writeClient = BigQueryServicesFactory.instance(connectOptions).storageWrite();
-            return writeClient.createStreamWriter(streamName, protoSchema, enableConnectionPool);
+            logger.debug("Creating BigQuery StreamWriter in subtask {}", subtaskId);
```
Comment (clmccart) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java:161]: i think it might be worth logging this at the info level. thoughts?

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java [hunk 2]
```diff
@@ -138,24 +150,43 @@ void addToAppendRequest(ByteString protoRow) {
...
+            streamWriter =
+                    writeClient.createStreamWriter(streamName, protoSchema, enableConnectionPool);
         } catch (IOException e) {
             logger.error("Unable to create StreamWriter for stream {}", streamName);
             throw new BigQueryConnectorException("Unable to create StreamWriter", e);
         }
     }
 
+    /** Creates a write stream and StreamWriter for appending to BigQuery table. */
+    void createWriteStreamAndStreamWriter(
```
Comment (clmccart) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java:174]: is there a simple way to refactor this so that it invokes createStreamWriter + some additional logic instead of duplicating code here?

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryDefaultWriter.java [hunk 3]
```diff
@@ -43,25 +42,33 @@
...
-    /** Accept record for writing to BigQuery table. */
+    /**
+     * Accept record for writing to BigQuery table.
+     *
+     * @param element Record to write
+     * @param context {@link Context} for input record
+     */
     @Override
     public void write(IN element, Context context) {
+        totalRecordsSeen++;
```
Comment (clmccart) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryDefaultWriter.java:71]: this might be out of scope for this PR but it might be worth trcking the element sizes here as well

Comment (clmccart) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryDefaultWriter.java:71]: ie, totalRecordsSeenSize

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/throttle/WriteStreamCreationThrottler.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    private final int writerId;
+
+    public WriteStreamCreationThrottler(int writerId) {
+        this.writerId = writerId;
+    }
+
+    public void throttle() {
+        int waitSeconds = writerId % MAX_BUCKETS;
+        try {
+            // Sleep does nothing if input is 0 or less.
```
Comment (clmccart) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/throttle/WriteStreamCreationThrottler.java:52]: we should probably log here that we're throttling and for how long

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/throttle/WriteStreamCreationThrottler.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ * correspond to a specific "wait" duration before calling the CreateWriteStream API.
+ *
+ * <p>Note that actual separation between CreateWriteStream invocations across all writers will not
+ * ensure exact QPS of 3, because neither all writers are initialized at the same instant, nor do
+ * they all identify the need to create a write stream after some uniform fixed duration. Given
+ * these uncontrolled variations, this throttler aims to achieve ~3 QPS on a best effort basis.
+ */
+public class WriteStreamCreationThrottler implements Throttler {
+
+    public static final int MAX_BUCKETS = BigQueryExactlyOnceSink.MAX_SINK_PARALLELISM / 3;
```
Comment (clmccart) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/throttle/WriteStreamCreationThrottler.java:41]: i dont see where BigQueryExactlyOnceSink.MAX_SINK_PARALLELISM is defined. looks like it's not apart of this PR

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/throttle/WriteStreamCreationThrottlerTest.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import java.time.Duration;
+import java.time.Instant;
+
+import static org.junit.Assert.assertTrue;
+
+/** Tests for {@link WriteStreamCreationThrottler}. */
+public class WriteStreamCreationThrottlerTest {
+
+    @Test
+    public void testThrottle() {
```
Comment (clmccart) [inline flink-1.17-connector-bigquery/flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/throttle/WriteStreamCreationThrottlerTest.java:30]: can we add a few more tests here for some of the edge cases (ie, weird subtaskid values, etc)?

**Review by clmccart — APPROVED** [review]

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/throttle/WriteStreamCreationThrottler.java (hunk 5 above)
Comment (jayehwhyehentee) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/throttle/WriteStreamCreationThrottler.java:41]: Its not.
https://github.com/GoogleCloudDataproc/flink-bigquery-connector/blob/8254a5e099405fbb6e009593c46569cea3c27dca/flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/BigQueryBaseSink.java#L39

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryDefaultWriter.java (hunk 3 above)
Comment (jayehwhyehentee) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BigQueryDefaultWriter.java:71]: @prashastia this should go into our document tracking Flink metrics

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java (hunk 1 above)
Comment (jayehwhyehentee) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java:161]: Done

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java (hunk 2 above)
Comment (jayehwhyehentee) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/writer/BaseWriter.java:174]: Yep, thanks

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/throttle/WriteStreamCreationThrottler.java (hunk 4 above)
Comment (jayehwhyehentee) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/throttle/WriteStreamCreationThrottler.java:52]: Done

File: flink-1.17-connector-bigquery/flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/throttle/WriteStreamCreationThrottlerTest.java (hunk 6 above)
Comment (jayehwhyehentee) [reply flink-1.17-connector-bigquery/flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/throttle/WriteStreamCreationThrottlerTest.java:30]: Done
